fdbserver/DataDistributionQueue.actor.cpp

```diff
@@ -1307,14 +1363,59 @@ ACTOR Future<Void> dataDistributionRelocator(DDQueueData* self,
 
 						bestTeams.emplace_back(bestTeam.first.get(), bestTeam.second);
 					}
+
+					// update physicalShardIDCandidate
+					if (CLIENT_KNOBS->SHARD_ENCODE_LOCATION_METADATA && tciIndex==0 && rd.dataMoveIdSetDelayed) { // && rd.dataMoveIdSetDelayed && foundTeams
+						ASSERT(foundTeams);
+						ShardsAffectedByTeamFailure::Team primaryTeam = 
+							ShardsAffectedByTeamFailure::Team(bestTeams[tciIndex].first->getServerIDs(), true);
+						if (forceToNewPhysicalShard) {
+							physicalShardIDCandidate = deterministicRandom()->randomUInt64();
+						} else {
+							physicalShardIDCandidate = self->shardsAffectedByTeamFailure->getPhysicalShardFor(primaryTeam);
+							if (physicalShardIDCandidate==UID().first()) { // if failed to find an available physicalShard
+								physicalShardIDCandidate = deterministicRandom()->randomUInt64();
+							}
+						}
+					}
+
 					tciIndex++;
 				}
+
+				if (CLIENT_KNOBS->SHARD_ENCODE_LOCATION_METADATA && rd.dataMoveIdSetDelayed && bestTeams.size()>1) {
+					if (!bestTeams[1].first->isHealthy()) {
+						foundTeams = false;
+						forceToNewPhysicalShard = true;
+					}
+				}
+
 				// once we've found healthy candidate teams, make sure they're not overloaded with outstanding moves
 				// already
 				anyDestOverloaded = !canLaunchDest(bestTeams, rd.priority, self->destBusymap);
 
 				if (foundTeams && anyHealthy && !anyDestOverloaded) {
 					ASSERT(rd.completeDests.empty());
+					if (CLIENT_KNOBS->SHARD_ENCODE_LOCATION_METADATA && rd.dataMoveIdSetDelayed) {
+						ASSERT(physicalShardIDCandidate!=UID().first());
+						rd.dataMoveId = UID(physicalShardIDCandidate, deterministicRandom()->randomUInt64());
+						auto inFlightRange = self->inFlight.rangeContaining(rd.keys.begin);
+						inFlightRange.value().dataMoveId = rd.dataMoveId;
+						ASSERT(inFlightRange.range() == rd.keys);
+						ASSERT(inFlightRange.value().randomId == rd.randomId);
+						ASSERT(inFlightRange.value().dataMoveId == rd.dataMoveId);
+						self->dataMoves.insert(rd.keys, DDQueueData::DDDataMove(rd.dataMoveId));
+					}
+					// update teamPhysicalShard mapping
+					if (CLIENT_KNOBS->SHARD_ENCODE_LOCATION_METADATA) {
```

Comment:
This is not necessary when we are restoring a data move, since the mapping is constructed during the initialization.

Reply:
For the case `rd.keys != ranges[r] && rd.isRestore()`, we don't have `rrs.dataMoveId` decided yet. We need decide a dataMoveId for this case here.